docket/common/utils.py

```diff
@@ -1,5 +1,252 @@
+        if i[0] == '.':
+            continue
+        if endswith and not i.endswith(endswith):
+            continue
+        if startswith and not i.startswith(startswith):
+            continue
+        file_path = os.path.join(path, i)
+        stats = os.stat(file_path)
+        if stats.st_size < 1:
+            continue
+        #if stats.st_mtime + 1 > epoch(datetime.utcnow()):
+        #    continue
+        files.append(i)
+    return files
+
+def file_modified(path, format=None):
+    """ return the modified datetime of the file at path
+        if format is specified - use strftime and return a string instead
+    """
+    t = from_epoch(os.path.getmtime(path))
+    if format:
+        return t.strftime(format)
+    return t
 
 def is_sequence(arg):
     return (not hasattr(arg, "strip") and
             hasattr(arg, "__getitem__") or
             hasattr(arg, "__iter__"))
+
+def recurse_update(a, b):
+    """ add everything in b to a
+        _overwrites_ existing values even if the new value is None
+        you might think deepcopy for this, but it can't merge two dictionaries
+    """
+    for k,v in b.items():
+        if isinstance(v, Mapping):
+            a[k] = recurse_update(a.get(k, {}), v)
+        else:
+            a[k] = v
+    return a
+
+def file_lock(f, timeout=2):
+    """ True if f was locked, False otherwise """
+    giveUp = datetime.utcnow() + timedelta(seconds=timeout)
+    while datetime.utcnow() < giveUp:
+        try:
+            flock(f, LOCK_EX | LOCK_NB)
+            return True
+        except IOError as e:
+            if e.errno == errno.EAGAIN:
+                time.sleep(0.1)
+    return False
+
+def update_yaml(path, val = None):
+    """ given a path to a yaml file:
+            if val: update val from the file, update the file, and return val.
+            else:   load the file and return that.
+        returns
+            False - locking failed, no update occurred
+            new value - File data, updated with provided value
+    """
+    if not path:
+        return False
+    locked = False
+
+    with open(path, 'a+b') as f:
+        if not file_lock(f, timeout=Config.get("LOCK_TIMEOUT", 2)):
+            Config.logger.error("Failed to lock {}".format(path))
+            return False
+
+        f.seek(0)
+        data = f.read()
+        if len(data):
+            tmp = yaml.load(data)
+            if not val:
+                Config.logger.debug("loaded {}".format(path))
+                # No changes to write.
+                return tmp
+            elif hasattr(val, 'update'):
+                Config.logger.debug("{}.update( {} )".format(val,tmp))
+                val.update(tmp)
+            elif hasattr(val, '__add__'):
+                Config.logger.debug("{} += {}".format(val,tmp))
+                val += tmp
+            else:
+                Config.logger.error("loaded {}, can't update {}".format(path,val))
+                return val
+
+        f.seek(0)
```

Comment:
I'd like to discuss this more, but a copy->write->move operation might be safer than the locking mechanism. It's possible this could end up on NFS.

Reply:
Redis is now the default meta-data store. 
A copy is saved to disk (still uses a lock) after the request is finalized. (FAIL, SUCCESS...)
Future improvements might include a 'active requests' queue in redis, which solves the problem: 'Docket isn't telling me anything about my queued request while it waits'